### ISSUE TITLE
[Bug] [NPU] Router GEMM output should always be fp32

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#34861]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

**The router GEMM returns bf16 on NPU.** The router GEMM output should always be fp32 in DeepSeek architectures.

`python/sglang/srt/models/deepseek_v2.py:545`

```python
elif not _is_cuda:
    logits = F.linear(hidden_states, self.weight, None)   # -> bf16
```

The prefill-CP branch above it (`deepseek_v2.py:527`) has the same shape — `linear_bf16_fp32` on CUDA, bare `F.linear` otherwise.

The CUDA branches of `MoEGate.forward` both return fp32: `dsv3_router_gemm(out_dtype=torch.float32)` (`deepseek_v2.py:538`) and `linear_bf16_fp32` (`deepseek_v2.py:548`).

The NPU routing path already expects fp32. `python/sglang/srt/hardware_backend/npu/moe/topk.py:55` and `:91` upcast both operands before the top-k:

```python
x=router_logits.to(torch.float32),
...
correction_bias.to(torch.float32) if correction_bias is not None else ...
```

but `.to(torch.float32)` cannot recover what `F.linear` already rounded to bf16, so the upcast is effective for the correction bias and cosmetic for the logits.

This is the same defect as #34857, which is confirmed to cause precision issues on the ROCm/aiter branch (`deepseek_v2.py:543`). The correction-bias half of #34857 does not apply here: the bf16 allocation at `deepseek_v2.py:481` is `_use_aiter`-gated, so on NPU the bias stays fp32 as stored in the checkpoint.

#29783 fixed the same defect on the CUDA side, switching the fallback from `F.linear` to `linear_bf16_fp32` so the output is fp32. The `not _is_cuda` branch was left on `F.linear`.

**Comment by b8zhong** [comment]
@ping1jing2 Can you take a look at this? As it affects the accuracy of NPU vs CUDA.

**Comment by b8zhong** [comment]
Hi @iforgetmyname, could you help take a look at this? Thank you very much.

**Comment by iforgetmyname** [comment]
> Hi [@iforgetmyname](https://github.com/iforgetmyname), could you help take a look at this? Thank you very much.

ok，i will take a look

**Comment by randgun** [comment]
Maybe convert input dtype to torch.float32 and then call torch.mm can solve your problem?
